### ISSUE TITLE
Improve support for C++ concepts

### ISSUE THREAD

**Issue by milasudril** [doxygen/doxygen#9569]
**Describe the bug**
As of v1.9.4, doxygen has basic support for concepts, however, it is not possible to document individual "members". Documentation comments on individual members is simply stripped from the documentation.

**Expected behavior**
Concepts should be presented in a similar way as abstract base classes.

**To Reproduce**
```c++
/**
 * \brief Defines the requirements of a "source"
 *
 * A source is something that can be read from, for example, an input buffer or a file.
 *
 * \ingroup de-serialization
 *
 */
template<class T>
concept source = requires(T a)
{
	/**
	 * \brief Shall try to consume next byte and return it. If it possible to consume another
	 *  byte, it shall return std::nullopt, which implies an End-Of-File condition.
	 */
	{ read_byte(a) } -> std::same_as<std::optional<char>>;
};
```
Becomes no more than

```c++
template<class T>
concept anon::source =  requires(T a)
    {
        
        { read_byte(a) } -> std::same_as<std::optional<char>>;
    }
```

**Version**
Version 1.9.4 build with libclang 14 on Ubuntu 22.04.

**Additional context**
The configuration file enables clang assisted parsing with -std=c++20 as option to clang.

**Comment by albert-github** [comment]
- What are you missing?
- What is in your doxygen settings file: so Can you please attach a, small, self contained example (source+configuration file in a, compressed, tar or zip file) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.
- I only see a `\ingroup` nut no definition of the group


@doxygen is it by design that in the code fragment the comment is missing even when `STRIP_CODE_COMMENTS = NO` ?

**Comment by milasudril** [comment]
I mean that it should write a more complete documentation page. Currently, it picks up the documentation of the concept itself, but it cannot produce a page which contains documentation about individual requirements. I think it can use the same format as a class, and document produce documentation for each requirement as if they were some kind of member function. Here I need to document that `read_byte` should consume next byte, and also, that std::nullopt is return in case of EOF, etc.

**Comment by h8672** [comment]
What happens if you surround it with **addtogroup** and surround with ``` @{ @} ```?
I cant add classes to a group with **ingroup** if it's not part of a group itself, because it would just create a module link to that group.
If the group has no name, I don't think it will be added.

I changed the code slightly, change the marking
```cpp
/**
 * \brief Defines the requirements of a "source"
 *
 * A source is something that can be read from, for example, an input buffer or a file.
 *
 * \addtogroup de-serialization
 * \{
 */
template<class T>
concept source = requires(T a)
{
	/**
	 * \brief Shall try to consume next byte and return it. If it possible to consume another
	 *  byte, it shall return std::nullopt, which implies an End-Of-File condition.
	 */
	{ read_byte(a) } -> std::same_as<std::optional<char>>;
};
/** \} */
```

**Comment by milasudril** [comment]
Apparently, it does not work for you, and yours is somewhat worse

* Concepts are no variables
* Concepts should not have an "initial value"

I want it to be formatted similar to other compounds (mostly like a if it were a class). Each requirement should be documented as a member. I also think this was scoped out for the initial support. The point of this issue is to improve on that.

**Comment by AlexPuchmayr** [comment]
I agree with the original author. The actual implementation is more or less copying the concept definition in pretty-formatted form, but there's no way to explain each individual requirement separately. What is the purpose of this requirement? If its a function, what is it expected to do? Error conditions?

**Comment by milasudril** [comment]
I played around with python-clang (version 15).

```python
import clang.cindex

def visit_nodes(node, src_file, depth):
    if node.location.file == None or node.location.file.name == src_file:
        for i in range(0, depth):
            print('    ', end = '')
        print('%s, %s, %s'%(node.kind, node.spelling, node.raw_comment))

        for child in node.get_children():
            visit_nodes(child, src_file, depth + 1)

def load_symbols(src_file, compiler_options = None):
    index = clang.cindex.Index.create()
    translation_unit = index.parse(src_file, compiler_options)
    visit_nodes(translation_unit.cursor, src_file, 0)

if __name__ == '__main__':
    import sys
    if len(sys.argv) < 2:
        exit(1)
    load_symbols(sys.argv[-1], sys.argv[1:-1])
```

And I got this:

```
CursorKind.TRANSLATION_UNIT, testdata/test.hpp, None
    CursorKind.CONCEPT_DECL, source, /**
 * \brief Defines the requirements of a "source"
 *
 * A source is something that can be read from, for example, an input buffer or a file.
 *
 * \ingroup de-serialization
 *
 */
        CursorKind.TEMPLATE_TYPE_PARAMETER, T, None
        CursorKind.REQUIRES_EXPR, , None
            CursorKind.PARM_DECL, a, None
                CursorKind.TYPE_REF, T, None
            CursorKind.DECL_REF_EXPR, , None
                CursorKind.OVERLOADED_DECL_REF, read_byte, None
            CursorKind.DECL_REF_EXPR, a, None
            CursorKind.NAMESPACE_REF, std, None
            CursorKind.TEMPLATE_REF, same_as, None
            CursorKind.NAMESPACE_REF, std, None
            CursorKind.TEMPLATE_REF, optional, None
```

It is clear that it does not return any documentation for the compound requirements, which definitely is an issue if you want to use libclang.

**Comment by albert-github** [comment]
From https://github.com/doxygen/doxygen/issues/9569#issuecomment-1488909076it is unclear, to me, what you have been doing
- which commands did you issue?
- what did you use as input?

**Comment by milasudril** [comment]
I mean that , if you enable clang assisted parsing (as said in the opening comment), Doxygen may not get any documentation from of individual requirements from libclang, and thus, there cannot be any documentation. Maybe this issue should be raised towards clang. As a side note, this AST would not be very easy to work with, since it does not give any end marker for a compound requirement.

**Comment by SimeonEhrig** [comment]
We have the same problem. We want to use C++ concepts to describe the interface of an object. What are the member types, what are the member function and what they are doing and more. But if I add body documentation, it is completely ignored if the HTML page is rendered.

Our use case is, that a function `void f(alpaka::concepts::impl::IMdSpan auto input)` is defined and the user should read the documentation of the concept `alpaka::concepts::impl::IMdSpan` to understand what he can do with the input. The user should not care about the actual object and it's class documentation, which is used to call the function.

Here is our source code (a little bit shortened and I tried different keywords):
```c++
         /** Interface concept which describes how to access multi-dimension host- and device side memory.
         * An alpaka::mdspan like object does not store any information about memory location, e.g. is the memory
         *located on a CPU or a GPU.
         *
         * @param t Object of the alpaka::mdspan like object. Can have a const modifier or not.
         * @param mut_t Mutable object of the alpaka::mdspan like object. Does not have the const modifier.
         * @param const_t Constant object of the alpaka::mdspan like object. Needs to have the const modifier.
         * @param vec Vector with the same number of elements like the dimension of the alpaka::mdspan like object. Is
         *used to call the access operator.
         *
         *
         **/
        template<typename T, typename MutT, typename ConstT>
        concept IMdSpan = requires(T t, MutT mut_t, ConstT const_t, alpaka::Vec<uint32_t, T::dim()> vec) {
            /// The element type. Can be const or not.
            typename T::value_type;
            /// @brief The element reference type, depending on T::value_type, it is const or not const.
            typename T::reference;
            /// The constant element reference type. Is const all the time.
            typename T::const_reference;
            typename T::pointer;
            typename T::const_pointer;
            typename T::index_type;

            requires std::copy_constructible<T>;
            requires std::assignable_from<MutT&, MutT&>;
            // TODO(SimeonEhrig): check if assignment fulfills const correctness
            // requires std::assignable_from<decltype(const_t)&, decltype(const_t)&>;
            requires std::movable<MutT>;

            { T::dim() } -> std::same_as<uint32_t>;
        };
```

Here is the HTML rendering:

<img width="1150" height="1057" alt="Image" src="https://github.com/user-attachments/assets/e78f0f70-de81-43a0-8c7c-335d60e71769" />

**Comment by albert-github** [comment]
@SimeonEhrig 

- which version of doxygen did you use (the current doxygen version is 1.14.0)

to be able to properly reproduce your issue
- Can you please attach a, small, self contained example (source+configuration file in a, compressed, tar or zip file!) that allows us to reproduce the problem?
   Please don't add external links as they might not be persistent (also references to GitHub repositories are considered non persistent).

**Comment by SimeonEhrig** [comment]
@albert-github I created a reproducer example with doxygen 1.14.

Therefore I generated a new configuration file with doxygen 1.14 and I only changed the `INPUT = ` configuration, that the source and header files are parsed.

[doxygen_concept.tar.gz](https://github.com/user-attachments/files/22508766/doxygen_concept.tar.gz)

I did some in body documentations of the `typename`s and the member function calls. In the rendered HTML, the documentation strings are missing. Instead, there are only empty lines.

<img width="736" height="927" alt="Image" src="https://github.com/user-attachments/assets/475fbb07-c342-4e4b-982f-c636745186f4" />

**Comment by albert-github** [comment]
The part that is circled is part of a source code listing and the setting `STRIP_CODE_COMMENTS` is set to `YES` so it is logical that we don't see the comment. It looks a bit strange that the `typename`s etc. are not show as it items concept (@doxygen reason? bug?)

**Comment by milasudril** [comment]
It should include documentation comments.

**Comment by albert-github** [comment]
When I understand your comment correctly you should use `STRIP_CODE_COMMENTS = NO` in your `Doxyfile`

**Comment by SimeonEhrig** [comment]
`STRIP_CODE_COMMENTS = NO` does not solve the actual problem. With the option, the comments becomes visible if I open the source code (`concepts.hpp`) but not in the rendered view, like the comment above: https://github.com/doxygen/doxygen/issues/9569#issuecomment-3327048370

With this workaround, the solution would be, to write in each doc string of each concept: `If you view the doxygen documentation in a web browser, please open the source code to get the documentation of the concept members`. Otherwise the user will never see the body documentation in the web view.

An ugly side effect would be, that the doc strings are also visible for classes and functions, where everything is working like expected and it is unnecessary.

**Comment by albert-github** [comment]
A bit late (sorry).

@doxygen looks like the results from the concepts are not optimal and some informations should be distributed to other places as well. Furthermore it looks like that the "members" in the concept and not mentioned with brief / detailed description (i.e. e.g. `value_type` or `size()` in the example). What do you think?

**Comment by doxygen** [comment]
@albert-github I agree that comments should be kept in the concept documentation independently of STRIP_CODE_COMMENTS. A concept itself doesn't really have members, it is just an expression which can also contain statements which can make it look a bit like a class. It could be represented as a tree, I guess, but I'm not sure if that would help with understandability. Ideally, the Doxygen comments for the expression parts are rendered in output format, along with the actual definition in plain text.

**Comment by doxygen** [comment]
@SimeonEhrig, @milasudril I've now let Doxygen keep any comments in the concept definition. As mentioned above, ideally, the comments should be rendered as Doxygen comments, so consider this a minimal implementation.

**Comment by SimeonEhrig** [comment]
It displays the comments now, which is much better.

<img width="1077" height="956" alt="Image" src="https://github.com/user-attachments/assets/0e760b8f-a523-4113-9f0c-08cae18cbbdf" />

> As mentioned above, ideally, the comments should be rendered as Doxygen comments, so consider this a minimal implementation.

I agree. This would be pretty helpful. But at least with have comments instead empty lines now :-) Thanks

**Comment by doxygen** [comment]
@milasudril @SimeonEhrig I've made some further improvements. The concept definition should now show internal Doxygen style comments with the correct markup. The only limitation is that the comments need to be on a separate line and not on the same line as the code. Let me know if you see any issues.

**Comment by SimeonEhrig** [comment]
@doxygen Thank you. I played a little bit with it.

```cpp
            /** typically the alignment of the value_type.
            * 
            * @details
            *
            * Calculating the alignment can be complicated.
            *     
            * @return An Alignment Object
            * 
            * @note Something important here
            */
           { t.getAlignment() } -> alpaka::concepts::Alignment;
```

<img width="1095" height="287" alt="Image" src="https://github.com/user-attachments/assets/e7c890dc-cccb-41f4-833c-500fc3936b56" />

It is already in a good shape :-)

A last feature, which I missing is the automatic parsing of `Public Types`, `Public Member Functions` and more like in a class. That would be pretty cool, if this would works. At the moment, we create the sections for concepts manually via doc string of the concept itself.
